### ISSUE TITLE
Dvorak keyboard layout disappeared from NixOS/Xorg

### ISSUE THREAD

**Issue by peti** [NixOS/nixpkgs#25526]
I have configured my keyboard layout via

    services.xserver.layout = "dvorak";

in `configuration.nix` for ages, but todays `nixos-rebuild` complained that this layout no longer exists:

~~~
building path(s) ‘/nix/store/p4na6ps7zd3agf8hc4rrl55h94msdh44-xkb-layouts-exist’

Some of the selected keyboard layouts do not exist:

  dvorak

Set services.xserver.layout to the name of an existing keyboard
layout (check /nix/store/dvxdqdsg8z9mfppam4drvhxpgy1vk8ir-xkeyboard-config-2.20/etc/X11/xkb/symbols for options).

builder for ‘/nix/store/5n01hrkp2hfb5mm1r5jmmn1j37fp86n2-xkb-layouts-exist.drv’ failed with exit code 255
~~~~

Alas, `/nix/store/dvxdqdsg8z9mfppam4drvhxpgy1vk8ir-xkeyboard-config-2.20/etc/X11/xkb/symbols` does not offer anything like "dvorak", so I now wonder what happened to this layout, why it is no longer available, and what I can do to get it back?

P.S.: I was trying to build nixos-unstable at c5badb123a.

**Comment by tohl** [comment]
services.xserver.layout has been broken since 17.03 for me.  I think
somebody already created an issue for this/something similar.

**Comment by 7c6f434c** [comment]
Hm, didn't know that would work… I thought that `dvorak` has to be specified as 
```
  services.xserver = {
    layout = "us";
    xkbVariant = "dvorak";
  };
```
when I was merging #23709 … Should it be reverted then?

The layout didn't really go away, but the check added assumes one specific way of choosing the layout (the one that is usually documented in relation to X).

@lheckemann

**Comment by lheckemann** [comment]
Sorry about that!

I'll have a look at getting the checker to work precisely like xkb does it. In the meantime, it could be reverted if it's acutely causing major issues, but given that at least for the dvorak case there's a workaround I don't think it's that bad? In any case, I'll have a look at fixing it this weekend.

**Comment by peti** [comment]
I've changed my configuration file to the style from https://github.com/NixOS/nixpkgs/issues/25526#issuecomment-299402708 and that works fine for the time being.

**Comment by aszlig** [comment]
This was broken since c5c0459a60c1e64afc66c3b76434b88961d67abb (#5638), so that check introduced there should probably be expanded to account for this.

**Comment by aszlig** [comment]
Oh, never mind, that's the commit @7c6f434c was already mentioning.

**Comment by aszlig** [comment]
Also, we should probably add the [keymap VM tests](https://github.com/NixOS/nixpkgs/blob/a35ec5dda688f6eedbb281671ee11261c769610e/nixos/tests/keymap.nix) to the `tested` jobset, to ensure something like this would block a channel update.

**Comment by aszlig** [comment]
Hm, this is a bit harder to achieve, because the layout is actually parsed from the rules file rather than from the files in the symbols directory:

```sh-session
$ strace -f -e trace=file setxkbmap dvorak
...ld blurb...
access("/home/aszlig/.Xauthority", R_OK) = 0
open("/home/aszlig/.Xauthority", O_RDONLY) = 4
open("./rules/evdev-C.lst", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("./rules/evdev.lst", O_RDONLY)     = -1 ENOENT (No such file or directory)
open("/nix/store/45p7lzf7ha1ax8wlxhl31v63k8k8hq5b-setxkbmap-1.3.1/share/X11/xkb/rules/evdev-C.lst", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/nix/store/45p7lzf7ha1ax8wlxhl31v63k8k8hq5b-setxkbmap-1.3.1/share/X11/xkb/rules/evdev.lst", O_RDONLY) = 4
open("/nix/store/45p7lzf7ha1ax8wlxhl31v63k8k8hq5b-setxkbmap-1.3.1/share/X11/xkb/rules/evdev-C", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/nix/store/45p7lzf7ha1ax8wlxhl31v63k8k8hq5b-setxkbmap-1.3.1/share/X11/xkb/rules/evdev", O_RDONLY) = 4
```

So in theory the most reliable way is to use `setxkbmap`, like this:

```bash
setxkbmap -print "$layout" | xkbcomp - - 3>&1 1>&2 2>&3 > /dev/null | sed -n -e '/^Error:/,/^[^\t ]/p'
```

If there are any errors this would print it to stdout, so checking for validity would be a matter of `[ -z "$output" ]`.

Unfortunately, this will require an X display and thus would only work if we'd run it inside an XVFB, which I think is a bit too much for a small check.

Related: https://unix.stackexchange.com/a/43984

So in theory we could also parse the rule files, not sure which one is the better option.

Maybe @lheckemann or @7c6f434c has a better idea?

Also note from the above `strace`: Users can have their own rules which are not necessarily available at the time we're checking this.

**Comment by 7c6f434c** [comment]
Re: user rules — given that the check is for the service, these are supposed to not be available.

Re: tests — that would also serve as the documentation for the parameter…

Maybe the check should just be dropped; the breakage it prevents is not that hard to recover from (X.org should stay in `us` layout and a single xkbmap is enough)

**Comment by lheckemann** [comment]
I'd say we have to reopen #5638 if we do remove it completely. The reason that this was added in the first place is that the naming is a bit confusing for some countries — e.g. it being uk in one place and gb in another for the United Kingdom layout (and I always forget which one!). Not having a clear message as to why it fails is a pain.

I think parsing the rules files would be better, and I wanted to look into that but haven't got very far yet. Could it be as simple as `grep -E "$layout"'\s*=' ${cfg.xkbDir}/rules/evdev` in addition to checking the symbols directory?

**Comment by 7c6f434c** [comment]
Well, there is also `us(dvorak)` notation that also works in principle.

**Comment by lheckemann** [comment]
Thanks @aszlig, sorry I never got around to fixing this.

**Comment by vcunat** [comment]
See https://github.com/NixOS/nixpkgs/commit/44c64fef16ed5#commitcomment-22813503 .  /cc @aszlig and @lheckemann for the blocker.

**Comment by roberth** [comment]
The test is failing and blocking the channel. Any ideas?

Here's the job

https://hydra.nixos.org/build/202787191#tabs-summary

Here's the log
[4gfnq33lz3i8g6qyq9a64491s984p8wz-vm-test-run-keymap-dvorak.drv.txt](https://github.com/NixOS/nixpkgs/files/10282601/4gfnq33lz3i8g6qyq9a64491s984p8wz-vm-test-run-keymap-dvorak.drv.txt)

Here's the channel blocker issue
 - https://github.com/NixOS/nixpkgs/issues/207181

**Comment by aij** [comment]
This seems to have regressed again

```
The value `dvorak' for keyboard layout is invalid.

Please check the definition in `services.xserver.layout'.

Detailed XKB compiler errors:

    Couldn't find file "symbols/dvorak" in include paths
    1 include paths searched:
        /nix/store/f0p3jc5rlc1lyd5wwgfwgywh6rd5prh6-xkeyboard-config-2.39/etc/X11/xkb
    3 include paths could not be added:
        /homeless-shelter/.config/xkb
        /homeless-shelter/.xkb
        /etc/xkb
    Abandoning symbols file "(unnamed)"
    Failed to compile xkb_symbols
    Failed to compile keymap
```

Not sure when exactly, but it was working as of  `c16abd6c722be95d53eef2bdd62cef5ecabd240a` and is broken in `9ca785644d067445a4aa749902b29ccef61f7476`

**Comment by vcunat** [comment]
FWIW we have a dvorak test that's still succeeding:
https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.keymap.dvorak.x86_64-linux

**Comment by aszlig** [comment]
We have a [keymap test](https://github.com/NixOS/nixpkgs/blob/4fe0dfd1ea244116d4b08b7e67fd61e718bf1ccf/nixos/tests/keymap.nix) for this, which used to test `services.xserver.layout = "dvorak";` but it was changed in 513143fe4ef0734483640945250b39d1732f7298 to use `layout = "us"; xkbVariant = "dvorak";`. I also reviewed that change but apparently didn't notice [the change](https://github.com/NixOS/nixpkgs/commit/513143fe4ef0734483640945250b39d1732f7298#diff-cdbad7bca2f5f739c8a8d3058f51db1f20596845a907817eed240e113015b387R148-R150) away from just `layout = "dvorak"` to `xkbVariant = "dvorak"`.

Nevertheless, I've bisected the commits you mentioned:

```sh-session
$ git bisect start 9ca785644d067445a4aa749902b29ccef61f7476 c16abd6c722be95d53eef2bdd62cef5ecabd240a
$ git bisect run sh -c 'nix-store -r "$(nix-store -qR $(nix-instantiate -E '\''(import ./nixos { configuration.services.xserver.enable = true; configuration.services.xserver.layout = "dvorak"; }).vm'\'') | grep xkb-validated.drv)"'
```

The result is 5a450bdd9f3d5f79b5a95c5c0905509516bd7ab7, which indicates that it was something that was changed upstream and indeed the upstream commit that removed the alias seems to be <https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/commit/470ad2cd8fea84d7210377161d86b31999bb5ea6>.

So in summary: It should have been `layout = "us"; xkbVariant = "dvorak";` in the first place and now that upstream has even deleted those old aliases shouldn't be used anymore. Since on our end we validate the XKB options using the upstream database, the error you're getting is valid and not a bug on our end.

We could however add such an alias on the NixOS option level, but I'm not so sure this is a good idea to add old baggage from `xkeyboard-config` on our end.
